### PR TITLE
Restructured workflows

### LINKED ISSUE #568: Actions are run twice for internally opened pull request
## Issue

In order to enable external contributions we had to add `on: pull_request` to our workflow. This will trigger a run of a PR opened from a fork, but in the case of an internal branch of the project the event `push` and `pull_request` trigger the workflow twice. 

We need to trigger our workflow only when: 
- push to main
- pushes to PR
- Create a PR (external contributions from forks)
- Tags 

Usage of composite actions could prove useful: https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action

## Solution 1

Partition our jobs and have a workflow per row as shown in the image below:

image

## Solution 2

Find the combination of trigger events that allows us to run our current workflow in the scenarios listed in the top section of the issue. Something along the lines of:

```yml
on:
  # Trigger the workflow on push or pull request,
  # but only for the main branch
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
  # Also trigger on page_build, as well as release created events
  page_build:
  release:
    types: # This configuration does not affect the page_build event above
      - created
```

For more information: https://docs.github.com/en/actions/reference/events-that-trigger-workflows

### PR DESCRIPTION
**Pull request by jpalvarezl**

### CODE REVIEW

File: .github/workflows/deploy_prod.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  deploy:
+    runs-on: ubuntu-latest
+    env:
+      DOCKERHUB_ORG: gnosispm
+      DOCKERHUB_PROJECT: safe-client-gateway
+    steps:
+      - name: Checkout branch
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 # checks out all branches and tags, necessary for versioning /about endpoint
```
Comment (fmrsabino) [inline .github/workflows/deploy_prod.yml:20]: How does this affect the docker build that happens on `bash scripts/deploy_docker.sh`

File: .github/workflows/deploy_prod.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+      - name: Dockerhub login
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USER }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+
+      - name: Deploy tag
+        run: bash scripts/deploy_docker.sh ${GITHUB_REF##*/}
```
Comment (fmrsabino) [inline .github/workflows/deploy_prod.yml:29]: One suggestion for this is to use `${{ github.event.release.tag_name }}` it represents the tagged version – `vX.Y.Z`

Can probably be done in a separate PR

**Review by fmrsabino — APPROVED** [review]

File: .github/workflows/deploy_prod.yml (hunk 2 above)
Comment (jpalvarezl) [reply .github/workflows/deploy_prod.yml:29]: that would be definitely more readable :+1:

File: .github/workflows/deploy_prod.yml (hunk 1 above)
Comment (jpalvarezl) [reply .github/workflows/deploy_prod.yml:20]: I think the reason I did this, was so that the script you mention would find the tag. I remember that with a regular checkout of the branch, tags were somehow not included. There is probably a nicer way to do this, as it sounds like a pretty common use case.

File: .github/workflows/deploy_prod.yml (hunk 2 above)
Comment (fmrsabino) [reply .github/workflows/deploy_prod.yml:29]: I think this is only available on events triggered by the type `release` (not 100% sure)

**Comment by jpalvarezl** [comment]
We decided to not move forward with this
